File: AppBox/ViewController/PreferencesViewController/EmailPreferencesViewController/EmailPreferencesViewController.m
//
//  EmailPreferencesViewController.m
//  AppBox
//
//  Created by Vineet Choudhary on 28/03/17.
//  Copyright © 2017 Developer Insider. All rights reserved.
//

#import "EmailPreferencesViewController.h"

@interface EmailPreferencesViewController ()

@end

@implementation EmailPreferencesViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    //set user emails and password
    [emailTextField setStringValue:[UserData userEmail]];
    [personalMessageTextField setStringValue:[UserData userMessage]];
}

- (BOOL)isValidEmailDetails {
    if (emailTextField.stringValue.length > 0 && [MailHandler isAllValidEmail:emailTextField.stringValue]) {
        return YES;
    } else {
        [MailHandler showInvalidEmailAddressAlert];
    }
    return NO;
}

//MARK: - Control Actions
- (IBAction)saveButtonTapped:(NSButton *)sender {
    [self.view.window makeFirstResponder:self.view];
	if (emailTextField.stringValue.isEmpty && personalMessageTextField.stringValue.isEmpty) {
		[UserData setUserEmail:emailTextField.stringValue];
		[UserData setUserMessage:personalMessageTextField.stringValue];
		[ABHudViewController showStatus:@"Details Cleared!" forSuccess:YES onView:self.view];
	} else if ([self isValidEmailDetails]) {
		[UserData setUserEmail:emailTextField.stringValue];
		[UserData setUserMessage:personalMessageTextField.stringValue];
		[ABHudViewController showStatus:@"Details Saved!" forSuccess:YES onView:self.view];
    }
}

- (IBAction)sendTestMailButtonTapped:(NSButton *)sender {
    [self.view.window makeFirstResponder:self.view];
    
    if (![self isValidEmailDetails]) {
        return;
    }
    
    //create a test ipa upload info for demo email
    IPAUploadInfo *ipaUploadInfo = [[IPAUploadInfo alloc] init];
    [ipaUploadInfo setName:@"TestApp"];
    [ipaUploadInfo setVersion:@"1.0"];
    [ipaUploadInfo setBuild:@"1"];
    [ipaUploadInfo setEmails: emailTextField.stringValue];
    [ipaUploadInfo setAppShortShareableURL:[NSURL URLWithString:@"https://getappbox.com"]];
    [ipaUploadInfo setPersonalMessage: [MailHandler parseMessage:personalMessageTextField.stringValue forIPAUploadInfo:ipaUploadInfo]];
    
    //send mail
    [ABHudViewController showStatus:@"Sending Test Mail" onView:self.view];
	weakify(self);
    [MailGun sendMailWithProject:ipaUploadInfo.abpProject complition:^(BOOL success, NSError *error) {
		dispatch_async(dispatch_get_main_queue(), ^{
			strongify(self);
			[ABHudViewController hideAllHudFromView:self.view after:0];
			if (success) {
				[ABHudViewController showStatus:@"Mail Sent" forSuccess:YES onView:self.view];
			} else {
				[ABHudViewController showStatus:@"Mail Failed" forSuccess:NO onView:self.view];
			}
		});
    }];
}

@end

File: AppBox/ViewController/PreferencesViewController/EmailPreferencesViewController/EmailPreferencesViewController.swift
//
//  EmailPreferencesViewController.swift
//  AppBox

import AppKit
import AppBoxCore
import os

public final class EmailPreferencesViewController: NSViewController {

    private static let log = Logger(subsystem: "com.developerinsider.AppBox", category: "EmailPreferences")
    private var model: EmailModel?

    public override func loadView() {
        let model = EmailModel(emails: UserData.userEmail(), message: UserData.userMessage())
        self.model = model
		model.onSave = { [weak self] in
			self?.saveTapped()
		}
		model.onSendTest = { [weak self] in
			self?.sendTestTapped()
		}
        view = EmailPreferencesHost.makeView(model: model)
    }

    // MARK: - Actions (invoked by the SwiftUI form's callbacks)

    private func isValidEmails(_ emails: String) -> Bool {
        if !emails.isEmpty, MailHandler.isAllValidEmail(emails) {
            return true
        }
        MailHandler.showInvalidEmailAddressAlert()
        return false
    }

    private func saveTapped() {
        guard let model else { return }
        let emails = model.emailsText
        let message = model.messageText
        let emailsEmpty = emails.trimmingCharacters(in: .whitespaces).isEmpty
        let messageEmpty = message.trimmingCharacters(in: .whitespaces).isEmpty
        if emailsEmpty && messageEmpty {
            UserData.setUserEmail(emails)
            UserData.setUserMessage(message)
            Common.showAlert(withTitle: "Saved", andMessage: "Email settings cleared.")
        } else if isValidEmails(emails) {
            UserData.setUserEmail(emails)
            UserData.setUserMessage(message)
            Common.showAlert(withTitle: "Saved", andMessage: "Email settings saved.")
        }
    }

    private func sendTestTapped() {
        guard let model else { return }
        let emails = model.emailsText
        let message = model.messageText
        guard isValidEmails(emails) else { return }
        guard DropboxSession.isAuthorized else {
            Common.showAlert(
				withTitle: "Connect Dropbox",
				andMessage: "Connect your Dropbox account first to send a test email.")
            return
        }

        var personalMessage: String?
        if !message.trimmingCharacters(in: .whitespaces).isEmpty {
            personalMessage = message
        }
        let request = BuildEmailRequest(
            name: "TestApp",
			version: "1.0",
			build: "1",
            to: emails.components(separatedBy: ",")
				.map { $0.trimmingCharacters(in: .whitespaces) }
				.filter { !$0.isEmpty },
            installURL: URL(string: "https://getappbox.com")!,
            personalMessage: personalMessage)

        model.setSendingTest(true)
        Task { [weak self] in
            do {
                try await AppServices.serviceClient.sendBuildEmail(request)
                await MainActor.run {
                    self?.model?.setSendingTest(false)
                    Common.showAlert(
						withTitle: "Mail Sent",
						andMessage: "The test email was sent successfully.")
                }
            } catch {
                await MainActor.run {
                    self?.model?.setSendingTest(false)
                    Self.log.error("Test mail failed: \(error.localizedDescription, privacy: .public)")
                    Common.showAlert(
						withTitle: "Mail Failed",
						andMessage: "Unable to send the test email. See the AppBox logs in Console.app for details.")
                }
            }
        }
    }
}
